### ISSUE TITLE
Allow loading Common.php function overrides

### ISSUE THREAD

**Issue by lonnieezell** [codeigniter4/CodeIgniter4#2101]
Originally discussed in this thread: https://forum.codeigniter.com/thread-74096.html

There is currently no elegant way for a developer to override any of the methods in Common.php. I think this can be remedied very simply: 

- In `system/bootstrap` check for the existence of a file, `app/Config/Functions.php` (or maybe `app/Functions.php`) prior to loading the Common.php file.
- If it exists load it in.

That's it. If a developer needs to override the function he can do it directly in that file or load other files that contain it, and not feel dirty editing `public/index.php`.

**Comment by jason-napolitano** [comment]
How about app/Config/Common.php or app/Common.php? for consistency across app and system paths. I may be able to find time in my busy day to produce a PR for this too, since I'd love to contribute to the framework I've used for eons =D

**Comment by jason-napolitano** [comment]
How about `app/Common.php`? For consistency across `app/` and `system/` paths. 

If the function exists in `app/Common.php`, use that and not the function inside of `system/Common.php`. This could be checked during the bootstrap process during framework execution. 

I may be able to find time in my busy day to produce a PR for this too, since I'd love to contribute to the framework I've used for eons.

**Comment by jason-napolitano** [comment]
How about shipping Codeigniter >= RC1 with a blank `app/Common.php`? This keeps it consistent across  the `app/ `and `system/` paths and allows developers to replace core functions with their own.

Basically, if the function exists in `app/Common.php`, use that function, and not the function inside of `system/Common.php`. This could be checked during the bootstrap process during framework execution. Think of a "master helper" file that they could replace system functions in, and create functions that may be used anywhere in their application's codebase.

I may be able to find time in my busy day to produce a PR for this too, since I'd love to contribute to the framework I've used for eons.

EDIT: If I may, here a simple solution without a PR. Consider it a proof of concept:

Replace:
https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/bootstrap.php#L99

```php
require_once SYSTEMPATH  .  'Common.php';
```

With
```php

if (file_exists(APPPATH . 'Common.php) {
    require_once APPPATH . 'Common.php';
}

require_once SYSTEMPATH . 'Common.php';
```

This way, if `app/Common.php` exists, those functions will get loaded first, then `system/Common.php`. Forcing the `if ( ! function_exists('...') )` checks in `system/Common.php` to fail, thus only loading functions from `app/Common.php`.

Sorry for the ramble, it is late and I am tired ha-ha. if you'd still like a PR, let me know =D

**Comment by jason-napolitano** [comment]
How about shipping Codeigniter >= RC1 with a blank/empty `app/Common.php`? This keeps it consistent across the `app/` and `system/` paths and allows developers to replace core functions with their own, paired with the ability to create procedural functions that may be used anywhere in their application's codebase.

Basically, if the function exists in `app/Common.php` use that function, and not the function inside of `system/Common.php`. This could be very simply checked with an `if()` during the bootstrap process. One way to look at it is in terms of a "master helper" file that they could replace system functions in.

I may be able to find time in my busy day to produce a PR for this too, since I'd love to contribute to the framework I've used for eons. 

For example, here a simple solution without a PR. Consider it a proof of concept:

Replace:
https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/bootstrap.php#L99

```php
require_once SYSTEMPATH  .  'Common.php';
```

With:
```php
// Make sure that the file exists. EG: That the dev's didn't delete it, etc.
if ( file_exists(APPPATH . 'Common.php) ) {

    // Require if it exists
    require_once APPPATH . 'Common.php';
}

// Require the Common.php file from within the framework
require_once SYSTEMPATH . 'Common.php';
```

This way, if `app/Common.php` exists, those functions will get loaded first, then `system/Common.php`. Forcing the `if ( ! function_exists('pre_existing_function') )` checks in `system/Common.php` to fail, thus only loading functions from `app/Common.php`.

Sorry if this is a bit of a ramble, it is late and I am tired ha-ha. If you'd like a PR, let me know =D

**Comment by jason-napolitano** [comment]
How about shipping Codeigniter >= RC1 with a blank/empty `app/Common.php`? This keeps the nomenclature consistent across the `app/` and `system/` paths and allows developers to replace core functions with their own, paired with the ability to create procedural functions that may be used anywhere in their application's codebase.

Basically, if the function exists in `app/Common.php` use that function, and not the function inside of `system/Common.php`. This could be very simply checked with an `if()` during the bootstrap process. One way to look at it is in terms of a "master helper" file that they could replace system functions in.

I may be able to find time in my busy day to produce a PR for this too, since I'd love to contribute to the framework I've used for eons. 

For example here a simple solution without a PR. Consider it a proof of concept:

Replace:
https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/bootstrap.php#L99

```php
require_once SYSTEMPATH  .  'Common.php';
```

With
```php
// Make sure that the file exists. EG: That the dev's didn't delete this file
if ( file_exists(APPPATH . 'Common.php) ) {
    // Require if it exists
    require_once APPPATH . 'Common.php';
}

// Require the Common.php file from within the framework
require_once SYSTEMPATH . 'Common.php';
```

This way, if `app/Common.php` exists, those functions will get loaded first, then `system/Common.php`. Forcing the `if ( ! function_exists('...') )` checks in `system/Common.php` to fail, thus only loading functions from `app/Common.php`.

Sorry if this is a bit of a ramble, it is late and I am tired ha-ha. If you'd like a PR, let me know =D

**Comment by lonnieezell** [comment]
I think that's the perfect way to handle it. It matches up with the blank BaseController we're shipping with to encourage it's use for adding common elements across all controllers, etc. 

If you've got time for a PR that would be awesome. Otherwise someone will get to it before RC1 ships (hopefully very soon).

**Comment by jason-napolitano** [comment]
Okay cool beans. If I've time after picking up my car from the shop tomorrow after work, I will most certainly submit a PR. Maybe not with tests, but logic and docs would be my goal.

Off-topic. I believe a better way to have handled the `BaseController` class would've been to ship Codeigniter 4 with an `app/Core/Controller.php` file, simply for nostalgic purposes, but that's neither here nor there =D. 

Thanks Lonnie and my apologies if I do not get to this by tomorrow evening.

**Comment by MGatner** [comment]
I could also send this over. @jason-napolitano let me know if you aren't going to get to it.

**Comment by jason-napolitano** [comment]
If you are faster than I am @MGatner then that'd be amazing. Just in case I don't get to it in a timely fashion =)

**Comment by jason-napolitano** [comment]
Here is a basic addition. No docs are yet provided because I am in no way use to the documentation library. my apologies for that.

PR: 2105

**Comment by jason-napolitano** [comment]
Pull request submitted. Sorry there are no docs for you. I am not at all familiar with the documentation library to build them in a timely fashion. However the logic and `app/Common.php` file is there. Hope it helps and please inform me of any further actions required by me, changes, etc.

**Comment by jason-napolitano** [comment]
@MGatner that is what I ended up doing for https://github.com/codeigniter4/CodeIgniter4/pull/2110 PR since it is a simple one. I'll get my ducks in order for any potential future PR's so I can rest assured that everything is setup right.

Thanks again!

**Comment by jim-parry** [comment]
Fixed by #2105 and #2162
